pkg/kubelet/kuberuntime/remote/remote_runtime.go

```diff
@@ -0,0 +1,345 @@
+		PodSandboxId:  &podSandBoxID,
+		Config:        config,
+		SandboxConfig: sandboxConfig,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return resp.GetContainerId(), nil
+}
+
+// StartContainer starts the container.
+func (r *RemoteRuntime) StartContainer(rawContainerID string) error {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	_, err := r.runtimeClient.StartContainer(ctx, &runtimeApi.StartContainerRequest{
+		ContainerId: &rawContainerID,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StopContainer stops a running container with a grace period (i.e., timeout).
+func (r *RemoteRuntime) StopContainer(rawContainerID string, timeout int64) error {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	_, err := r.runtimeClient.StopContainer(ctx, &runtimeApi.StopContainerRequest{
+		ContainerId: &rawContainerID,
+		Timeout:     &timeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveContainer removes the container. If the container is running, the container
+// should be force removed.
+func (r *RemoteRuntime) RemoveContainer(rawContainerID string) error {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	_, err := r.runtimeClient.RemoveContainer(ctx, &runtimeApi.RemoveContainerRequest{
+		ContainerId: &rawContainerID,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ListContainers lists all containers by filters.
+func (r *RemoteRuntime) ListContainers(filter *runtimeApi.ContainerFilter) ([]*runtimeApi.Container, error) {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	resp, err := r.runtimeClient.ListContainers(ctx, &runtimeApi.ListContainersRequest{
+		Filter: filter,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Containers, nil
+}
+
+// ContainerStatus returns the container status.
+func (r *RemoteRuntime) ContainerStatus(rawContainerID string) (*runtimeApi.ContainerStatus, error) {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	resp, err := r.runtimeClient.ContainerStatus(ctx, &runtimeApi.ContainerStatusRequest{
+		ContainerId: &rawContainerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Status, nil
+}
+
+// Exec execute a command in the container.
+func (r *RemoteRuntime) Exec(rawContainerID string, cmd []string, tty bool, stdin io.Reader, stdout, stderr io.WriteCloser) error {
+	ctx, cancel := r.getContext()
+	defer cancel()
+
+	stream, err := r.runtimeClient.Exec(ctx)
+	if err != nil {
+		glog.Errorf("Get remote runtime client stream failed: %v", err)
+		return err
+	}
+
+	request := &runtimeApi.ExecRequest{
+		ContainerId: &rawContainerID,
+		Cmd:         cmd,
+		Tty:         &tty,
+	}
+	err = stream.Send(request)
+	if err != nil {
+		glog.Errorf("Send exec request to remote runtime failed: %v", err)
+		return err
+	}
+
+	errChanOut := make(chan error, 1)
+	errChanIn := make(chan error, 1)
+	exit := make(chan bool)
+
+	go func(stdout, stderr io.WriteCloser) {
+		for {
+			defer close(errChanOut)
```

Comment:
This doesn't seem right to me, e.g. the second time the loop tries to reads the stdout/stderr, the defer is called again, which will probably cause a close on a closed channel later.